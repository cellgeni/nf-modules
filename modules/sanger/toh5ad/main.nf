process TOH5AD {
    tag "Converting ${meta.id}'s file to .h5ad"
    label 'process_single'
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'docker://quay.io/cellgeni/metacells-python:latest':
        'quay.io/cellgeni/metacells-python:latest' }"

    input:
    tuple val(meta), path(input, name: 'input/*')
    val delimiter

    output:
    tuple val(meta), path("*.h5ad"), emit: h5ad
    path "versions.yml"                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    convert_to_h5ad.py \
            --input ${input} \
            --sample_id ${meta.id} \
            ${delimiter ? "--delimiter ${delimiter}" : ""} \
            ${args} \
            --output ${prefix}.h5ad

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        anndata: \$( python -c "import anndata; print(anndata.__version__)" )
        scanpy: \$( python -c "import scanpy; print(scanpy.__version__)" )
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.h5ad

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        anndata: \$( python -c "import anndata; print(anndata.__version__)" )
        scanpy: \$( python -c "import scanpy; print(scanpy.__version__)" )
    END_VERSIONS
    """
}
