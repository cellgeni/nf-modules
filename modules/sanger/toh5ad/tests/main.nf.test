nextflow_process {

    name "Test Process TOH5AD"
    script "../main.nf"
    process "TOH5AD"

    tag "local/toh5ad"

    test("Conversion of .h5 file to .h5ad") {
        when {
            process {
                """
                input[0] = [[id: 'pbmc_1k_h5'], file(params.test_data_base + "toh5ad/pbmc_1k.h5")]
                input[1] = "___"
                """
            }
        }

        then {
            // basec run health
            assert process.success
            assert process.exitStatus == 0

            // h5ad files
            assert process.out.h5ad
            assert process.out.h5ad.get(0).get(0).id ==~ "pbmc_1k_h5"
            assert process.out.h5ad.get(0).get(1) ==~ ".*/pbmc_1k_h5.h5ad"

            // versions.yml
            assert process.out.versions
            assert process.out.versions.get(0) ==~ ".*/versions.yml"

            // Snapshot all output channels of a process
            assert snapshot(process.out).match()

        }
        
    }

    test("Conversion of 10x .mtx file to .h5ad") {
        when {
            process {
                """
                input[0] = [[id: 'pbmc_1k_mtx'], file(params.test_data_base + "toh5ad/filtered_feature_bc_matrix")]
                input[1] = "___"
                """
            }
        }

        then {
            // basec run health
            assert process.success
            assert process.exitStatus == 0

            // h5ad files
            assert process.out.h5ad
            assert process.out.h5ad.get(0).get(0).id ==~ "pbmc_1k_mtx"
            assert process.out.h5ad.get(0).get(1) ==~ ".*/pbmc_1k_mtx.h5ad"

            // versions.yml
            assert process.out.versions
            assert process.out.versions.get(0) ==~ ".*/versions.yml"

            // Snapshot all output channels of a process
            assert snapshot(process.out).match()

        }
        
    }

    test("Run -stub for .mtx and .h5 files") {

        options "-stub"
        when {
            process {
                """
                input[0] = Channel.fromList([
                    [[id: 'pbmc_1k_h5'], file(params.test_data_base + "toh5ad/pbmc_1k.h5")],
                    [[id: 'pbmc_1k_mtx'], file(params.test_data_base + "toh5ad/filtered_feature_bc_matrix")]
                ])
                input[1] = Channel.fromList(["___", "___"])
                """
            }
        }

        then {
            // basec run health
            assert process.success
            assert process.exitStatus == 0
            assert process.trace.succeeded().size() == 2

            // h5ad files
            assert process.out.h5ad
            assert process.out.h5ad.size() == 2

            // versions.yml
            assert process.out.versions
            with(process.out.versions) {
                assert size() == 2
                assert get(0) ==~ ".*/versions.yml"
                assert get(1) ==~ ".*/versions.yml"
            }

            // Snapshot all output channels of a process
            assert snapshot(process.out).match()

        }
        
    }

}