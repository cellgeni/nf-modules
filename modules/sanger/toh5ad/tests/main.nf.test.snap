{
    "Conversion of 10x .mtx file to .h5ad": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "pbmc_1k_mtx"
                        },
                        "pbmc_1k_mtx.h5ad:md5,07a06e51127967ec3d1b2e443e4326f0"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ],
                "h5ad": [
                    [
                        {
                            "id": "pbmc_1k_mtx"
                        },
                        "pbmc_1k_mtx.h5ad:md5,07a06e51127967ec3d1b2e443e4326f0"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T12:52:22.770619"
    },
    "Run -stub for .mtx and .h5 files": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "pbmc_1k_h5"
                        },
                        "pbmc_1k_h5.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "id": "pbmc_1k_mtx"
                        },
                        "pbmc_1k_mtx.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89",
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ],
                "h5ad": [
                    [
                        {
                            "id": "pbmc_1k_h5"
                        },
                        "pbmc_1k_h5.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        {
                            "id": "pbmc_1k_mtx"
                        },
                        "pbmc_1k_mtx.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89",
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T12:47:31.909586"
    },
    "Conversion of .h5 file to .h5ad": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "pbmc_1k_h5"
                        },
                        "pbmc_1k_h5.h5ad:md5,be44e2060e1bd1d56b7328ed48c2e7bb"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ],
                "h5ad": [
                    [
                        {
                            "id": "pbmc_1k_h5"
                        },
                        "pbmc_1k_h5.h5ad:md5,be44e2060e1bd1d56b7328ed48c2e7bb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc7171bc35713269ee1b51e99bb31b89"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-05-28T12:52:11.98103"
    }
}